src/sage/rings/species.py

```diff
@@ -0,0 +1,1752 @@
+                    new = P.zero()
+                    for tau, _ in taus:
+                        F = libgap.Intersection(libgap.ConjugateGroup(H, tau), G)
+                        new += P(PermutationGroup(gap_group=F, domain=L.domain()), dpart)
+                    res += c * d * new
+
+            return res
+
+        def _compose_with_weighted_singletons(self, names, multiplicities, degrees):
+            r"""
+            Compute the composition with
+            `(\sum_j m_{1,j} X_{1,j}, \sum_j m_{2,j} X_{2,j}, \dots)`
+            in the specified degrees.
+
+            The `k`-sort species ``self`` should be homogeneous.
+
+            INPUT:
+
+                - ``names``, the (flat) list of names of the result
+
+                - ``multiplicities``, a (flat) list of constants
+
+                - ``degrees``, a `k`-tuple of compositions `c_1,
+                  \dots, c_k`, such that the size of `c_i` is the
+                  degree of self in sort `i`.
+
+            EXAMPLES:
+
+            Equation (2.5.41)::
+
+                sage: P = PolynomialSpecies(QQ, ["X"])
+                sage: E2 = P(SymmetricGroup(2))
+                sage: E2._compose_with_weighted_singletons(["X"], [-1], [[2]])
+                -E_2 + X^2
+
+                sage: C4 = P(CyclicPermutationGroup(4))
+                sage: C4._compose_with_weighted_singletons(["X"], [-1], [[4]])
+                -C_4 + {((1,2)(3,4),)}
+
+            Exercise (2.5.17)::
+
+                sage: C4._compose_with_weighted_singletons(["X", "Y"], [1, 1], [[2, 2]])
+                E_2(XY) + X^2*Y^2
+                sage: C4._compose_with_weighted_singletons(["X", "Y"], [1, 1], [[3, 1]])
+                X^3*Y
+                sage: C4._compose_with_weighted_singletons(["X", "Y"], [1, 1], [[4, 0]])
+                C_4(X)
+
+            Auger et al., Equation (4.60)::
+
+                sage: C4._compose_with_weighted_singletons(["X", "Y"], [1, -1], [[2, 2]])
+                -E_2(XY) + 2*X^2*Y^2
+
+            TESTS::
+
+                sage: (C4+E2^2)._compose_with_weighted_singletons(["X"], [-1], [[4]])
+                -C_4 + {((1,2)(3,4),)} + E_2^2 - 2*X^2*E_2 + X^4
+
+            """
+            P = self.parent()
+            if not self.support():
+                return P.zero()
+            if not self.is_homogeneous():
+                raise ValueError("element is not homogeneous")
+
+            left = sum(c * M._compose_with_singletons(P.base_ring(),
+                                                      names,
+                                                      degrees)
+                       for M, c in self)
+            P = left.parent()
+            right = P.exponential(multiplicities,
+                                  list(chain.from_iterable(degrees)))
+            return left.hadamard_product(right)
+
+        def __call__(self, *args):
+            """
+
+            EXAMPLES::
+
+                sage: P = PolynomialSpecies(QQ, ["X"])
```

Comment:
E2(X^2) fails if we use P = PolynomialSpecies(ZZ, ["X"]) instead.

Reply:
Yes. This is unfortunate, but a consequence of the way I implement call

Reply:
sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
sage: P(PermutationGroup([[(1,2),(3,4)]]), {1: [1,2], 2: [3,4]})
E_2(X)*E_2(Y)

This should be E_2(X*Y), right? E_2(X)*E_2(Y) should be P(PermutationGroup([[(1,2)],[(3,4)]]), {1: [1,2], 2: [3,4]})